src/storage/src/table/state_table.rs

```diff
@@ -113,171 +116,109 @@ impl<S: StateStore> StateTable<S> {
-    ) -> StorageResult<StateTableRowIter<'a, S>> {
-        let mut cell_based_streaming_iter =
+    ) {
+        let mut cell_based_table_iter =
             CellBasedTableStreamingIter::new(keyspace, table_descs, epoch).await?;
         let pk_serializer = OrderedRowSerializer::new(order_types_vec.to_vec());
-        let cell_based_item = cell_based_streaming_iter.next().await.map_err(err)?;
-        let state_table_iter = Self {
-            mem_table_iter,
-            cell_based_streaming_iter,
-            cell_based_item,
-            pk_serializer,
-        };
-
-        Ok(state_table_iter)
-    }
 
-    /// This function scans kv pairs from the `shared_storage`(`cell_based_table`) and
-    /// memory(`mem_table`). If a record exist in both `cell_based_table` and `mem_table`, result in
-    /// `mem_table` is returned according to the operation(RowOp) on it. This is because the data in
-    /// memory is fresher.
-    pub async fn next(&mut self) -> StorageResult<Option<Row>> {
+        let mut cell_based_table_item = cell_based_table_iter.next().await.map_err(err)?;
+        let mut mem_table_item = mem_table_iter.next();
         loop {
-            let next_flag;
-            let res;
-            let cell_based_item = self.cell_based_item.take();
-            match (cell_based_item, self.mem_table_iter.peek()) {
+            match (&cell_based_table_item, &mem_table_item) {
                 (None, None) => {
-                    next_flag = NextOutcome::End;
-                    res = None;
+                    yield None;
                 }
                 (Some((_, row)), None) => {
-                    next_flag = NextOutcome::Storage;
-                    res = Some(row);
+                    yield Some(row.clone());
+                    cell_based_table_item = cell_based_table_iter.next().await.map_err(err)?;
                 }
                 (None, Some((_, row_op))) => {
-                    next_flag = NextOutcome::MemTable;
                     match row_op {
-                        RowOp::Insert(row) => {
-                            res = Some(row.clone());
-                        }
-                        RowOp::Delete(_) => {
-                            res = None;
-                        }
-                        RowOp::Update((_, new_row)) => {
-                            res = Some(new_row.clone());
+                        RowOp::Insert(row) | RowOp::Update((_, row)) => {
+                            yield Some(row.clone());
                         }
+                        _ => {}
                     }
+                    mem_table_item = mem_table_iter.next();
                 }
                 (Some((cell_based_pk, cell_based_row)), Some((mem_table_pk, mem_table_row_op))) => {
                     let mem_table_pk_bytes =
-                        serialize_pk(mem_table_pk, &self.pk_serializer).map_err(err)?;
+                        serialize_pk(mem_table_pk, &pk_serializer).map_err(err)?;
```

Comment:
One thing I'd like to optimize here is to serialize pk for only one time. Currently if we find cell based pk always < mem table pk, the pk will be serialized for multiple times. We can wrap the memtable iterator with something like `.map(|(k, v)| serialize_pk(k), v)` or wrap another async stream over it, so as to reduce the serialization overhead.

Reply:
Add a helper function to serialize once only.

`.map` seems to return a `&Result` that I cannot clone and I cannot `?` on `&Result`.

Wrap another async stream seems to add a unnecessary `await`?  I don't know whether the compiler may optimize it away.